### PR TITLE
Light FFI for the interpreter

### PR DESCRIPTION
**Pull request by xymus**
Intro an FFI for the interpreter. This is a first draft/proof of concept to support the light FFI and few more features, more work is needed to implement the full FFI and support for other languages.

When the interpreter has to execute an extern method, it compiles all the extern code of the modules into an external native shared library (.so). Then the interpreter load dynamically the module, and invoke the C implementation function. The handle to the library is cached so a module isn't compiled twice per execution.

The current version is limited to the light FFI with a few more features. So there is extern methods, native type support and extern classes but it does not support callbacks.

The interpreter can now run 7 out of the 24 tests for the FFI with C.

This version was kept "simple" by design, it lacks some not-so-hard to implement features as well as portability concerns:
* Choosing the compiler (hard coded gcc) and the compile folder.
* Recompile the extern library only when the source is updated.
* Clear the compile folder.
* Support shared library format on OS X.
* Ensure call compatibility between the interpreter and extern libraries

The remaining features of the FFI, the callbacks and the other languages, require a decent amount of work but less than this PR.

Note that not a single line of code was added to the FFI folder, it uses the exact same backend as the compiler without modifications.

The design was inspired by other interpreters and VM-based languages. The JNI expect the user to manually compile the extern code into a native shared library before executing the Java code. This FFI automates the same process.

---

Please ignore the first 6 commits from #1441 and #1442.

### CODE REVIEW

**Comment by jpages** [comment]
I found the design very smart and elegant but the code is hard to read (as expected, we can feel the pain)

By curiosity, you have an idea of the perfs of the solution ?

+1

**Comment by xymus** [comment]
I wouldn't know how to test the perfs... but here are some thoughts.

The compilation part is quite simple, it extracts the user code into C files with only glue code and the entry functions that are artificial. There is no Nit to C compilation and there is usually very little C code within a Nit module compared to the result of the Nit to C compilation. So both generating the C code and the execution of gcc is pretty fast.

The calls to the .so file could be made faster, at the cost of the simplicity of the implementation. Other interpreter languages (Ruby, Java, etc.) do not use a standardized entry function and instead call the implementation directly. The difference is that they must implement calls to all possible signatures of a C functions. That means to declare statically, in the C code of the interpreter, all the possible unpack and calls to functions with 0 to 32+ parameters of about 6 primitives types. They often use macro magic to do so, but it is not pretty. With our standardized entry functions we loose a small bit of performance but it will be easier to maintain.

Considering the perfs of our interpreter at this point, there is not much doubt in my mind that implementing performance critical code in C will be faster than in Nit. As usual, running small programs, even with the FFI, is much faster with the interpreter than with the compiler.

**Comment by xymus** [comment]
There are still a few problems:
* curl and all its users had compilation error because of callbacks in module headers, they are fixed by #1445.
* `is_address_null` is implemented as an intern and returns almost always false. It causes a lot of segfaults, of course the only protection against it always says that it is not null. But this would be solved by the next point. 
* This PR does not support the legacy NI. So many more programs will run on niti only once the legacy NI code has been updated to light FFI. Then we can also dump all the intern implementation of extern methods.
* The implementation of `NativeString` with a `FlatBuffer` does not have the information for `address_is_null`. This causes infinite loop on reading files. Either we reimplement it with native strings, as it would be much easier now. Or, and there's a real need for this, we differenciate between `NativeString` and `CString`. gcc really doesn't like the new C type `unsigned char*` for `NativeString`.

**Comment by xymus** [comment]
Rerolled as per @jpages comments and to fix a few more issues.

File: src/nitni/nitni_base.nit [hunk 1]
```diff
@@ -71,14 +71,15 @@ redef class MType
...
-	fun cname_blind: String is abstract
+	fun cname_blind: String do return "struct nitni_instance *"
 
 	# Representation of this type in mangled C
 	#   Object -> Object
 	#   Pointer -> Pointer
 	fun mangled_cname: String is abstract
 
-	# Does this types has a primitive reprensentation
+	# Does this types has a primitive representation?
```
Comment (R4PaSs) [inline src/nitni/nitni_base.nit]: :s/types has/type have

Comment (R4PaSs) [inline src/nitni/nitni_base.nit]: Just saw this is part of #1442, feel free to fix it or not since the PR is already ok will merge

**Comment by R4PaSs** [comment]
:shipit: !

**Comment by Morriar** [comment]
+1

**Comment by MehdiAit** [comment]
:shipit: !!!

**Comment by privat** [comment]
Impressive.
Since the light FFI seems only to be required for the dynamic compilation, can you interpret the interpreter that executes some basic ffi programs. E.g.

~~~sh
$ nit nit.nit tests/test_ffi_c_simple.nit
~~~

**Comment by xymus** [comment]
Rebased on the most recent master to resolve conflict, didn't change anything in the PR.

**Comment by privat** [comment]
+1

**Comment by xymus** [comment]
Rerolled to name the generated `.so` file with the module `c_name`.

**Comment by privat** [comment]
Tried to merge but it seems to broke the boostrap http://gresil.org/jenkins/job/nit_bootstrap/852/console
More investigation is needed.

**Comment by xymus** [comment]
I brough back the hard coded -lm in the compiler, the bootstrap should be OK now.

**Comment by privat** [comment]
ok, will retry this night

**Comment by privat** [comment]
Did you try to run tests?

~~~sh
$ cd tests
$ ./testfull --engine niti
~~~

On my machine I have some failures

fail:  complete_simple_android mpi_simple neo_doxygen_dump_args1 neo_doxygen_dump_args2 neo_doxygen_dump_args3 neo_doxygen_dump_args4 neo_doxygen_dump_args5 neo_doxygen_dump_args6 neo_doxygen_dump_args7 neo_doxygen_dump_args8 neo_doxygen_dump_args9 nitiwiki_args1 nitiwiki_args2 test_binary test_binary_alt1 test_binary_alt2 test_binary_alt3 test_c test_c_alt1 test_c_alt2 test_c_alt3 test_c_alt4 test_jvm test_jvm_alt1 test_jvm_alt2 test_jvm_alt3 test_jvm_alt4

The fast CI_github job do not run the tests with niti, it is why they where not detected before.
Usually, regression if engines are catched by the big tests. However since I catched them now, I will not stack it for the merge as is.

For instance

~~~diff
--- sav/test_c.res	2015-06-12 13:49:57.967783056 -0400
+++ out/test_c.res	2015-06-12 14:59:27.591693436 -0400
@@ -1,7 +1 @@
-0
-0
-1
-2
-3
-4
-01234
+Runtime error: Abstract method `is_cprimitive` called on `MVirtualType` (nitni/nitni_base.nit:81)
~~~

or

~~~diff
--- sav/nitiwiki_args1.res	2015-06-08 20:44:02.781900122 -0400
+++ out/nitiwiki_args1.res	2015-06-12 14:55:49.511126003 -0400
@@ -1,10 +1,2 @@
-nitiWiki
-name: wiki2
-config: ../contrib/nitiwiki/tests/wiki1/config2.ini
-url: http://localhost/
-
-There is modified files:
- + pages
- + /pages/index.md
-
-Use nitiwiki --render to render modified files
+Caught signal : Segmentation fault
+Segmentation fault
~~~

or

~~~diff
--- sav/test_jvm.res	2015-06-08 20:44:02.793899887 -0400
+++ out/test_jvm.res	2015-06-12 15:00:09.359035209 -0400
@@ -1,17 +1,12 @@
+sh: 1: JAVA_HOME: not found
+sh: 1: JAVA_HOME: not found
+sh: 1: JAVA_HOME: not found
+sh: 1: JAVA_HOME: not found
+In file included from nit_compile/jvm._ffi.c:7:0:
+../lib/jvm.nit:27:18: fatal error: jni.h: No such file or directory
+  #include <jni.h>
+                  ^
+compilation terminated.
+Runtime error: FFI Error: Failed to compile C code using `gcc -Wall -c -fPIC -I nit_compile -g -o nit_compile/jvm._ffi.o nit_compile/jvm._ffi.c -I $(JAVA_HOME)/include/ -I $(JAVA_HOME)/include/linux/` (../lib/jvm.nit:79)
 Compilation des classes Java ...
 Initialisation de la JVM ...
----------------------Test 1----------------------
-From java, pushing premier
-From java, pushing deuxi?me
-From java, pushing troisi?me
-From java, popping premier
-premier
-From java, popping deuxi?me
-deuxième
-From java, popping troisi?me
-troisième
---------------------Test 2---------------------
-true
-D
-10
-15.5
~~~

File: src/interpreter/dynamic_loading_ffi/dynamic_loading_ffi.nit [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	fun pointer=(value: Pointer) `{ self->value_Pointer = value; `}
+
+	# The `Instance` held by this cell
+	fun instance: Instance `{ return (Instance)self->value_Pointer; `}
+
+	# The `Instance` held by this cell
+	fun instance=(value: Instance) `{ self->value_Pointer = value; `}
+
+	# The `NativeString` held by this cell
+	fun native_string: NativeString `{ return (unsigned char*)self->value_Pointer; `}
```
Comment (privat) [inline src/interpreter/dynamic_loading_ffi/dynamic_loading_ffi.nit]: This will cause some issue since #1502 is now merged

**Comment by xymus** [comment]
There's a few errors aded by the latest commits that I can easily fix, but the rest takes much more work:
* The segfaults are expected since the interpreter always returns false on `Pointer::address_is_null`. The implementation in the lib is in legacy NI so I cannot remove the hack at this point. (this includes neo_doxygen, nitiwiki and nitdoc)
* `test_binary` fails on the hack for the PrimitiveNativeFile.
* `test_jvm` fails because its ldflags are destined to a Makefile. I was actually planning to generate a Makefile to create the .so library, it should take care of this problem.

Juste from this analysis, there is a lot of clean up to do before all these tests pass:
* Remove `PrimitiveNativeFile` and the intern implementation of `NativeString`. Since I expect these changes to be controversial I kept them for a future PR.
* Update some critical methods like `is_address_null` to use the light FFI and remove the hackish intern implementation.

So @privat, I see 2 solutions, which do you prefer?
1. We keep this PR open until everything is fixed, which imply waiting for at least a partial update of the lib to the light FFI.
2. We blacklist those tests for the FFI for the time being. They were not supported previously anyway.

**Comment by privat** [comment]
I propose to blacklist them in a smart way: with UNDEFINED set in the `tests/sav/nit/fixme/*.res` files of tests that fails.
The benefit of `UNDEFINED` in `fixme` sub-directory, is that if one day one of the blacklisted tests become functional, ./tests.sh will notify us.

**Comment by privat** [comment]
still some test failed http://gresil.org/jenkins/job/nit_tests/engine=niti/899/testReport/ :(

**Comment by xymus** [comment]
Rerolled to fix remaining tests.

**Comment by xymus** [comment]
Now it should be OK!

**Pull request merged**